lib/spack/spack/variant.py

```diff
@@ -22,17 +22,552 @@
+            return MultiValuedVariant
+        elif self.values == (True, False):
+            return BoolValuedVariant
+        return SingleValuedVariant
+
+
+@lang.key_ordering
+class MultiValuedVariant(object):
+
+    @staticmethod
+    def from_node_dict(name, value):
+        if isinstance(value, list):
+            value = ','.join(value)
+            return MultiValuedVariant(name, value)
+        elif str(value).upper() == 'TRUE' or str(value).upper() == 'FALSE':
+            return BoolValuedVariant(name, value)
+        return SingleValuedVariant(name, value)
+
+    def __init__(self, name, value):
+        self.name = name
+
+        # Stores 'value' after a bit of massaging
+        # done by the property setter
+        self._value = None
+        self._original_value = None
+
+        # Invokes property setter
+        self.value = value
+
+    @property
+    def value(self):
+        """Returns a tuple of strings containing the values stored in
+        the variant.
+
+        :return: a tuple of strings
+        :rtype: tuple
+        """
+        return self._value
+
+    @value.setter
+    def value(self, value):
+        self._value_setter(value)
+
+    def _value_setter(self, value):
+        # Store the original value
+        self._original_value = value
+
+        # Store a tuple of CSV string representations
+        # Tuple is necessary here instead of list because the
+        # values need to be hashed
+        f = StringIO(str(value))
+        try:
+            t = next(csv.reader(f, skipinitialspace=True))
+        except StopIteration:
+            t = []
+
+        # With multi-value variants it is necessary
+        # to remove duplicates and give an order
+        # to a set
+        self._value = tuple(sorted(set(t)))
+
+    def _cmp_key(self):
+        return self.name, self.value
+
+    def copy(self):
+        """Returns an instance of a variant equivalent to self
+
+        :return: a copy of self
+        :rtype: any variant
+
+        >>> a = MultiValuedVariant('foo', True)
+        >>> b = a.copy()
+        >>> assert a == b
+        >>> assert a is not b
+        """
+        return type(self)(self.name, self._original_value)
+
+    def satisfies(self, other):
+        """Returns true if other.name == self.name and other.value is
+        a strict subset of self. Does not try to validate.
+
+        :param other: constraint to be met for the method to
+            return True
+        :return: True or False
+        :rtype: bool
+        """
+        # If types are different the constraint is not satisfied
+        if type(other) != type(self):
+            return False
+
+        # If names are different then `self` does not satisfy `other`
+        # (`foo=bar` does not satisfy `baz=bar`)
+        if other.name != self.name:
+            return False
+
+        # Otherwise we want all the values in `other` to be also in `self`
+        return all(v in self.value for v in other.value)
```

Comment:
Exactly. Besides we are not really growing asymptotically in input size: I don't think we'll ever exceed 10 elements for `self.value` (famous last words).